src/ai/backend/manager/idle.py

```diff
@@ -307,24 +305,133 @@ class UtilizationIdleChecker(BaseIdleChecker):
+        log.debug(f"ExecutionIdleChecker._disable_timeout({session_id})")
+        await self._redis.set(
+            f"session.{session_id}.last_access", "0", exist=self._redis.SET_IF_EXIST
+        )
+
+    async def _update_timeout(self, session_id: SessionId) -> None:
+        log.debug(f"ExecutionIdleChecker._update_timeout({session_id})")
+        t = await self._redis.time()
+        await self._redis.set(
+            f"session.{session_id}.last_access",
+            f"{t:.06f}",
+            expire=max(86400, int(self.idle_timeout.total_seconds() * 2)),
+        )
+
+    async def _session_started_cb(
+        self,
+        context: None,
+        source: AgentId,
+        event: SessionStartedEvent,
+    ) -> None:
+        await self._update_timeout(event.session_id)
+
+    async def _execution_started_cb(
+        self,
+        context: None,
+        source: AgentId,
+        event: ExecutionStartedEvent,
+    ) -> None:
+        await self._disable_timeout(event.session_id)
+
+    async def _execution_exited_cb(
+        self,
+        context: None,
+        source: AgentId,
+        event: ExecutionFinishedEvent | ExecutionTimeoutEvent | ExecutionCancelledEvent,
+    ) -> None:
+        await self._update_timeout(event.session_id)
+
+    async def _do_idle_check(
+        self,
+        context: None,
+        source: AgentId,
+        event: DoIdleCheckEvent,
+    ) -> None:
+
+        try:
+            return await super()._do_idle_check(context, source, event)
+        finally:
+            pass
+
+    async def check_session(self, session: Row, dbconn: SAConnection) -> bool:
+
+        redis = await aioredis.create_redis("redis://127.0.0.1:8111/")
+
+        redis_out = await redis.get(str(session["session_id"]), encoding=None)
+        try:
+            live_stat = msgpack.unpackb(redis_out)
+        except Exception:
+            return True
+
+        cpu_util_pct = float(live_stat["cpu_util"]["pct"])
+        mem_util_pct = float(live_stat["mem"]["pct"])
+
+        interval = self.timer.interval
+
+        if "m" in self.wind:
+            wind = int(self.wind.replace("m", ""))
+            wind_size = (wind * 60) / interval
+        elif "h" in self.wind:
+            wind = int(self.wind.replace("h", ""))
+            wind_size = (wind * 60 * 60) / interval
+        else:
+            raise Exception("Undefined unit of window interval")
+
+        if len(self.cpu_util_series) < wind_size:
+            self.cpu_util_series.append(float(cpu_util_pct))
+            self.mem_util_series.append(float(mem_util_pct))
+            print(self.cpu_util_series, len(self.cpu_util_series))
+            return True
+        else:
+            avg_cpu_util = sum(self.cpu_util_series) / len(self.cpu_util_series)
+            avg_mem_util = sum(self.mem_util_series) / len(self.mem_util_series)
+
+            self.cpu_util_series = []
+            self.mem_util_series = []
```

Comment:
Just clear the collected time-series utilization data when they overflow the window size? Wouldn't it be natural to pop out the oldest element from the data and append the latest one?